### ISSUE TITLE
Biometric authentication setup isn't working in Firefox

### ISSUE THREAD

**Issue by tangowithfoxtrot** [bitwarden/browser#2258]
### Steps To Reproduce

Filing on behalf of a user who has performed the following steps:
 
1. Installed the Bitwarden desktop application and Firefox extension.
2. Installed [Microsoft Visual C++ Redistributable](https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads).
3. Followed the [documented](https://bitwarden.com/help/article/biometrics/#enable-unlock-with-biometrics) setup process to configure biometric authentication.
4. Ensured that the Firefox extension has permission to interact with NativeMessaging: 
![permission](https://user-images.githubusercontent.com/5676771/148287496-8531c827-819c-42bc-9d3a-d964ccccfad9.png)

### Expected Result

Biometric authentication to be enabled.

### Actual Result

Biometric authentication is not enabled. Instead, they get this prompt:
![image](https://user-images.githubusercontent.com/5676771/148287681-b6ca04fc-942f-43e7-b132-b3261bb4666a.png)

### Screenshots or Videos

![image](https://user-images.githubusercontent.com/5676771/148287753-b37015bb-a21f-4f79-865b-d169349bc9e6.png)

### Additional Context

Biometric authentication was enabled successfully for the user in Edge Chromium 96.0.1054.62 x64

### Operating System

Windows

### Operating System Version

Windows 11 Enterprise 21H2

### Web Browser

Firefox

### Browser Version

95.0.2

### Build Version

1.55.0

**Comment by Naxterra** [comment]
Additional test steps are below.

- Tested Firefox versions: 95.0.1 from Microsoft Store and 95.0.2 standalone.
- Deactivated and reactivated Windows Hello integration in the desktop app.
- Created a new profile in Firefox and installed only BW extension.
- Tried with and without VPN.
- Reinstalling OS didn't change the outcome either.
- OS is enrolled in Azure AD.

**Comment by Hinton** [comment]
Hi @Naxterra,

Could you check if the following registry key exists? Which can be done by opening the Registry Editor using `regedit.exe` in the run prompt. And pasting it in the address bar.

`HKEY_CURRENT_USER\Software\Mozilla\NativeMessagingHosts\com.8bit.bitwarden`

If it doesn't exist could you check if `HKEY_CURRENT_USER\Software\Mozilla\` exists?

**Comment by Naxterra** [comment]
> Hi @Naxterra,
> 
> Could you check if the following registry key exists? Which can be done by opening the Registry Editor using `regedit.exe` in the run prompt. And pasting it in the address bar.
> 
> `HKEY_CURRENT_USER\Software\Mozilla\NativeMessagingHosts\com.8bit.bitwarden`
> 
> If it doesn't exist could you check if `HKEY_CURRENT_USER\Software\Mozilla` exists?

Hello

Mozilla exists but com.8bit.bitwarden doesn't exist.

![image](https://user-images.githubusercontent.com/41871948/150764937-641ecabd-daca-4c23-b84d-abef656a7819.png)

**Comment by Hinton** [comment]
@Naxterra can you try to disable browser integration, and re-enable it again? And check if the registry key gets created?

![image](https://user-images.githubusercontent.com/137855/150766701-7642d566-6dcf-4c86-8792-247808bbce6b.png)

**Comment by Naxterra** [comment]
> @Naxterra can you try to disable browser integration, and re-enable it again? And check if the registry key gets created?
> 
> ![image](https://user-images.githubusercontent.com/137855/150766701-7642d566-6dcf-4c86-8792-247808bbce6b.png)

I did that and it didn't create any entries.

**Comment by Hinton** [comment]
Hmm interesting.

Could you close the Bitwarden desktop application completely. Open a new powershell or cmd terminal, and start the application from there which should show some debug information. Afterwards paste the log here, and I can see if there's any information why it doesn't work.

It seems odd that Edge would work since it also relies on registry entries (although under the chrome namespace).

It's possible to manually create the registry entry.

```
key: com.8bit.bitwarden
value: C:\Users\<username>\AppData\Roaming\Bitwarden\browsers\firefox.json
```

**Comment by Naxterra** [comment]
> Hmm interesting.
> 
> Could you close the Bitwarden desktop application completely. Open a new powershell or cmd terminal, and start the application from there which should show some debug information. Afterwards paste the log here, and I can see if there's any information why it doesn't work.
> 
> It seems odd that Edge would work since it also relies on registry entries (although under the chrome namespace).
> 
> It's possible to manually create the registry entry.
> 
> ```
> key: com.8bit.bitwarden
> value: C:\Users\<username>\AppData\Roaming\Bitwarden\browsers\firefox.json
> ```

It worked after I created entry manually.

![image](https://user-images.githubusercontent.com/41871948/150770393-c4e6d440-ca85-42db-8789-cb59b4231b76.png)


![image](https://user-images.githubusercontent.com/41871948/150770329-f9320bbc-9446-421f-a370-f30fe256554d.png)

**Comment by Naxterra** [comment]
Solution should be Bitwarden must create these entries when installing the app. Users shouldn't create any registry entries manually

**Comment by Hinton** [comment]
@Naxterra we do, that's why I wanted the console output since obviously something went wrong.

Code responsible for creating the registry entries are identical between firefox and chrome.
https://github.com/bitwarden/desktop/blob/master/src/main/nativeMessaging.main.ts#L106-L110

**Comment by Naxterra** [comment]
But Edge is empty and Native Messaging policy is not set. https://docs.microsoft.com/en-us/deployedge/microsoft-edge-policies#nativemessaginguserlevelhosts

![image](https://user-images.githubusercontent.com/41871948/150773488-c165379b-cfea-4d74-8501-b2ea70a54fee.png)

**Comment by Hinton** [comment]
Edge fallsback to Chromium, like most chromium forks.

**Comment by 0neac** [comment]
hi, i just have the same problem. 
I follow your checks : same results i have the firefox key but not the bitwarden's one. So i create it manually. 
Now, the extension poses me another problem. To active the biometry it wait the confirmation of the desktop one but i receive nothing on it. Have you an idea ? 


![Screenshot_6](https://user-images.githubusercontent.com/100795863/156421758-e44c9243-c589-425b-bac6-71a61d117c42.png)

**Comment by BananenRitter** [comment]
Hi guys i got my extension on firefox 102.0.1 (64-Bit) finally to work!!!
1. i go the program via bitwardens website not the ms store

2. i reinstalled the extension and enabled/renabled the permissions in the app the which didnt work. the problem was that even tho i enabled it, i got a message that said it wasnt enabled in the desktop program.

3. there was no entry in HKEY_CURRENT_USER\Software\Mozilla\NativeMessagingHosts\com.8bit.bitwarden
i manually added key: com.8bit.bitwarden , value: C:\Users\<username>\AppData\Roaming\Bitwarden\browsers\firefox.json
like you suggested then i restarted my pc.

4. opened bitwaren then firefox with admin rights

5.. enabled/renabled the permissions again but it didnt work either.

6. then i was mad and enabled all taskbar icon settings as well as "verify browser integration via code", set the permission under "manage extentions" again and tried to enable biometrics in the extention and now it magically worked. then i deactivated all the settings but not the first taskbar one because i dont wanted them and it still works (the language is german) 
![grafik](https://user-images.githubusercontent.com/100603352/178164280-731651d6-a781-4fca-989e-d24e9e4b7733.png)
![grafik](https://user-images.githubusercontent.com/100603352/178164338-eb3e5add-b932-45c2-a092-00fc1513c1d0.png)
So i guess one of these settings is required for the first use of this feature.  Maybe it does activate something.

Ive also noticed that in the registry there was one entry before i added the one youve suggested.
under Computer\HKEY_CURRENT_USER\SOFTWARE\Mozilla\NativeMessagingHosts
was com.robin.messagehost
with the value C:\Program Files (x86)\Power Automate Desktop\PAD.FirefoxManifest.json
which seemed odd to me because i manually installed windows power automate some time ago, i never really used it before but i am sure that bitwarden as well as the extention should work without installing this unrelated program right?

If i remember correct i needed to install a different Visual Studio file while installing power automate. Correct me if i am wrong but i think i read that these additional visual studio c++ files are related to DLL files which are exssential for certain system programs to work. And i recently saw a video that some programers edit these files for their programs to work which can break others because they rely on the original values. So my guess would be that this breaks the path of creating a certain file or path for forwarding which causes com.8bit.bitwarden not to be created.
Can you confirm that you had the same value in com.robin.messagehost?
Do you have power automate installed as well?

(i am not a programmer i just try to get it to work but i hope im still helpful with this information. If not just tell me to delete it i dont want to spam here :)

**Comment by MrBretticus** [comment]
> Hi guys i got my extension on firefox 102.0.1 (64-Bit) finally to work!!!
> 
> Version 2022.6.1 Shell 16.2.7 Renderer 96.0.4664.174 Node 16.9.1 Architecture x64 Edition Windows 10 Home Version 21H2 Betriebssystembuild 19044.1766 Leistung Windows Feature Experience Pack 120.2212.4180.0 2022-06 Kumulatives Update für Windows 10 Version 21H2 für x64-basierte Systeme (KB5014666)
> 
>     1. i go the program via bitwardens website not the ms store
> 
>     2. i reinstalled the extension and enabled/renabled the permissions in the app the which didnt work. the problem was that even tho i enabled it, i got a message that said it wasnt enabled in the desktop program.
> 
>     3. I installed Visual Studio X86 as suggested on community.bitwarden
> 
>     4. there was no entry in HKEY_CURRENT_USER\Software\Mozilla\NativeMessagingHosts\com.8bit.bitwarden
>        i manually added key: com.8bit.bitwarden , value: C:\Users<username>\AppData\Roaming\Bitwarden\browsers\firefox.json
>        like you suggested then i restarted my pc.
> 
>     5. opened bitwaren then firefox with admin rights
> 
> 
> 6.. enabled/renabled the permissions again but it didnt work either.
> 
>     7. then i was mad and enabled all taskbar icon settings as well as "verify browser integration via code", set the permission under "manage extentions" again and tried to enable biometrics in the extention and now it magically worked. then i deactivated all the settings but not the first taskbar one because i dont wanted them and it still works (the language is german)
>        ![grafik](https://user-images.githubusercontent.com/100603352/178164280-731651d6-a781-4fca-989e-d24e9e4b7733.png)
>        ![grafik](https://user-images.githubusercontent.com/100603352/178164338-eb3e5add-b932-45c2-a092-00fc1513c1d0.png)
>        So i guess one of these settings is required for the first use of this feature.  Maybe it does activate something.
> 
> 
> Ive also noticed that in the registry there was one entry before i added the one youve suggested. under Computer\HKEY_CURRENT_USER\SOFTWARE\Mozilla\NativeMessagingHosts was com.robin.messagehost with the value C:\Program Files (x86)\Power Automate Desktop\PAD.FirefoxManifest.json which seemed odd to me because i manually installed windows power automate some time ago, i never really used it before but i am sure that bitwarden as well as the extention should work without installing this unrelated program right?
> 
> If i remember correct i needed to install a different Visual Studio file while installing power automate. Correct me if i am wrong but i think i read that these additional visual studio c++ files are related to DLL files which are exssential for certain system programs to work. And i recently saw a video that some programers edit these files for their programs to work which can break others because they rely on the original values. So my guess would be that this breaks the path of creating a certain file or path for forwarding which causes com.8bit.bitwarden not to be created. Can you confirm that you had the same value in com.robin.messagehost? Do you have power automate installed as well?
> 
> (i am not a programmer i just try to get it to work but i hope im still helpful with this information. If not just tell me to delete it i dont want to spam here :)

I can confirm that this also worked for me after enabling all the settings. I did not have to install anything else.

**Comment by serin113** [comment]
I've been experiencing this issue as well: the [manual fix mentioned above](https://github.com/bitwarden/clients/issues/2258#issuecomment-1019969507) worked fine, right until the "Allow browser integration" checkbox is unchecked and checked again.

Unchecking the option deletes the registry key as expected (nothing appears in the console log), but rechecking it fails to recreate the registry key, with the following console log:
```
Server path not specified, so defaulting to ipc.config.socketRoot + ipc.config.appspace + ipc.config.id /tmp/app.bitwarden
starting server on  /tmp/app.bitwarden
starting TLS server false
starting server as Unix || Windows Socket
{"HKLM\\SOFTWARE\\Mozilla\\Firefox":
12:57:26.532 > Not finding registry HKLM\SOFTWARE\Mozilla\Firefox skipping.
```

Note that the registry key here is looking in **HKLM** (HKEY_LOCAL_MACHINE) and not **HKCU** (HKEY_CURRENT_USER) as mentioned in the above link.

TL:DR; Bitwarden is writing to a nonexistent key.
